server/src/main/java/org/elasticsearch/common/util/ByteArray.java

```diff
@@ -51,6 +52,17 @@ static ByteArray readFrom(StreamInput in) throws IOException {
      */
     void fill(long fromIndex, long toIndex, byte value);
 
+    /**
+     * Fills this ByteArray with bytes from the given input stream
+     */
+    void fillWith(StreamInput in) throws IOException;
```

Comment:
We can use the BytesRef iterator instead of introducing a new method, but it seems unsafe since we can materialize the returned BytesRef. We have an assertion, but not a hard failure for checking that.